Adds workflow to build a demo

bishtgautam opened the pull request:
- Adds Github action to run two demos (`swe_roe/ex1.c` and `swe_roe/ex2.c`).
- Fixes `swe_roe/ex2.c` to run in parallel.
- Introduces an auxiliary DM with 1 DOF that is used to create a PETSc Vec that identifies
  if a grid cell is building or not.

Fixes #9

bishtgautam committed:
Adds workflow to build a demo

bishtgautam committed:
Run the ex1

bishtgautam committed:
Prints similiar message as ex1.c

bishtgautam committed:
Fixes the code for parallel run

bishtgautam committed:
Saves vectors in natural order

bishtgautam committed:
Corrects sn/cn values

bishtgautam committed:
Moving the definition of sn/cn within internal or boundary cases

bishtgautam committed:
Adss suffix to filename to indiciate number of procs

bishtgautam committed:
fixes format

bishtgautam committed:
Adds ex2 to Github actions

bishtgautam commented:
@jeff-cohere In this PR, I'm using for Github Action to [run](https://github.com/RDycore/toy-problems/actions/runs/3697798663/jobs/6263220366) the two demos (`ex1.c` and `ex2.c`) using the PETSc container used by the rdycore. The demos are currently running serially. Is there a way to run the demos in parallel by modifying code at https://github.com/RDycore/toy-problems/pull/11/files#diff-71b0abc932ce3afa499e28fe2386860a406f028eb953560619bae6b67a21114aR34?

jeff-cohere commented:
Hey Gautam. By "run the demos in parallel," do you mean run each demo sequentially using more than one process, or do you mean running them both at the same time? If the former, you just need to add `mpiexec -n <NPROC>` to the beginning of the commands. If you want to run both tests at the same time, I'm not sure we can guarantee that it will work well, since the free test runners provided by GitHub don't have very good hardware.

bishtgautam committed:
adding a demo to run in parallel

bishtgautam commented:
DAH, I just had to use `mpiexec -n <NPROC>`. I was overthinking how to figure out the `mpiexec` that ~was used to build~ will be compatible with PETSc.

bishtgautam committed:
Uses a new DM for creating a Vec for building

A new DM, `auxdm`,  created by cloning the original multi-DOF DM
and creating a section with 1 DOF.

bishtgautam committed:
fixes the format

jeff-cohere commented on swe_roe/ex2.c:
```diff
@@ -719,6 +719,8 @@ struct _n_RDyApp {
   char filename[PETSC_MAX_PATH_LEN];
   /// PETSc grid
   DM dm;
+  /// A DM for creating PETSc Vecs with 1 DOF
```
Maybe it would be good to think of these two DMs in terms of primary/secondary or prognostic/diagnostic variables. It's true that the "primary" DM holds the grid, which is shared between the two DMs, but I think a DM is really a set of DOF and their related function spaces.

This is just an FYI--I don't think anything is needed here unless you wanted to add a comment or change how you're referring to the DMs.

bishtgautam replied:
I agree that we should use better variable names for the RDycore lib, but current names are good enough for this repo.
